jetty-client/src/main/java/org/eclipse/jetty/client/HttpChannel.java

```diff
@@ -27,11 +27,13 @@
     protected static final Logger LOG = Log.getLogger(HttpChannel.class);
 
     private final HttpDestination _destination;
+    private final TimeoutCompleteListener _totalTimeout;
```

Comment:
the problem with having the TimeoutCompleteListener on the HttpChannel, is that for HTTP/2, the channel is not recycled, so we lose the benefit of the CyclicTimeout.

It is logically the right place, but only if the HttpChannel is recycled!

Also, we probably need to add a destroy method to HttpChannel, so that a busy HTTP/2 server does not end up with many dead HttpChannels being held in memory by their CyclicTimers

Reply:
@sbordet why did we turn off HttpChannel recycling in h2?  Does it still need to be off?

Reply:
@sbordet cancel these comments.  We do recycle the HttpChannel for the HttpClient, so all good with this PR (but we should review the server side default).

I think this PR LGTM to merge

Reply:
@gregw I don't see where `HttpChannel` is recycled for `HttpClient`. Your first concern is good IMHO: since we don't recycle `HttpChannel`, we don't get the benefits of `CyclicTimeout`.

So perhaps we really need to rethink the implementation by having the `CyclicTimeout` at the connection level; for HTTP/1.1 this will map 1-to-1 with a channel, and the connection is pooled; for HTTP/2 we will use a shared `CyclicTimeout` for all the channels created by a connection.

Reply:
Or perhaps we recycle HttpChannels?  They are recycled for HTTP/1 (`HttpChannelOverHTTP` is a [final field of HttpConnectionOverHTTP](https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-client/src/main/java/org/eclipse/jetty/client/http/HttpConnectionOverHTTP.java#L50)) so why not for HTTP/2?  We never really discovered a problem, we just suspected it.

The main components of the HttpChannel are the Sender and Receiver, both of which are designed to be reused.  Can you at least look up the issues we had with recycling and see?

The `CyclicTimeout` really matches well with the `HttpChannel`, as we need one timeout per active request and each channel can have only 1 active request.  It is the right spot for it.